### PR TITLE
📊 isd, gw: number of countries in each region

### PR DESCRIPTION
**Pull request by lucasrodes**
Add two datasets with the number of states/countries in each region.

Preview the dataset:

- [States in regions (Gleditsch)](http://staging-site-lucas/admin/datasets/6213)
- [States in regions (ISD)](http://staging-site-lucas/admin/datasets/6214)

~Before reviewing this PR, take a look at (only your assigned files):~

Waiting for merge:
- https://github.com/owid/etl/pull/1701

### CODE REVIEW

File: etl/steps/data/garden/countries/2023-09-25/gleditsch.py [hunk 1]
```diff
@@ -15,20 +26,83 @@ def run(dest_dir: str) -> None:
...
+
+    # Get numbers for World
+    tb_world = tb.groupby(["year"], as_index=False).agg({"id": "nunique"}).rename(columns={"id": "number_countries"})
+    tb_world["region"] = "World"
+
+    # Combine
+    tb_regions = pr.concat([tb_regions, tb_world], ignore_index=True)
+
+    # Add short name
+    tb_regions.metadata.short_name = f"{tb_regions.m.short_name}_regions"
```
Comment (paarriagadap) [inline etl/steps/data/garden/countries/2023-09-25/gleditsch.py:82]: The short_name can be included in the previous function

File: etl/steps/data/garden/countries/2023-09-25/shared.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
+from datetime import datetime as dt
+
+import owid.catalog.processing as pr
+import pandas as pd
+from owid.catalog import Table
+
+# Only for table tb_regions:
+# Latest year to have had a 31st of December
+LAST_YEAR = 2022  # Update to extend it further in time
```
Comment (paarriagadap) [inline etl/steps/data/garden/countries/2023-09-25/shared.py:9]: Change it to 2023 if you want to change it to this year

File: etl/steps/data/garden/countries/2023-09-25/gleditsch.meta.yml [hunk 3]
```diff
@@ -1,20 +1,31 @@
 # NOTE: To learn more about the fields, hover over their names.
 
-
 # Learn more about the available fields:
 # http://docs.owid.io/projects/etl/en/latest/architecture/metadata/reference/dataset/
```
Comment (paarriagadap) [inline etl/steps/data/garden/countries/2023-09-25/gleditsch.meta.yml:4]: Links don't work

File: etl/steps/data/garden/countries/2023-09-25/isd.meta.yml [hunk 4]
```diff
@@ -4,12 +4,33 @@
 # Learn more about the available fields:
```
Comment (paarriagadap) [inline etl/steps/data/garden/countries/2023-09-25/isd.meta.yml:4]: "Settlement" is mispelled in `startsettle` (can't comment there)

File: etl/steps/data/garden/countries/2023-09-25/isd.py [hunk 5]
```diff
@@ -15,25 +29,153 @@ def run(dest_dir: str) -> None:
...
+    log.info("isd: creating table with countries in region")
+    tb_regions = create_table_countries_in_region(tb=tb)
+
+    # Add to tables list
+    tables = [
+        tb.set_index(["cownum", "start", "end"], verify_integrity=True).sort_index(),
+        tb_regions.set_index(["region", "year"], verify_integrity=True).sort_index(),
+    ]
+
     # tb = tb.set_index(["country", "year"], verify_integrity=True)
```
Comment (paarriagadap) [inline etl/steps/data/garden/countries/2023-09-25/isd.py:58]: Delete comment

File: etl/steps/data/garden/countries/2023-09-25/isd.py [hunk 6]
```diff
@@ -15,25 +29,153 @@ def run(dest_dir: str) -> None:
...
+
+    # Get region name, then obtain number of countries per region per year
+    tb["region"] = tb["cownum"].apply(code_to_region)
+    # Get number of countries per region per year
+    tb_regions = (
+        tb.groupby(["region", "year"], as_index=False)
+        .agg({"cownum": "nunique"})
+        .rename(columns={"cownum": "number_countries"})
+    )
+    # Same as before, but with an alternate region set. Basically, instead of (Africa, Middle East) -> (Sub-Saharan Africa, North Africa & Middle East)
```
Comment (paarriagadap) [inline etl/steps/data/garden/countries/2023-09-25/isd.py:125]: If not being used, delete this comment

File: etl/steps/data/garden/countries/2023-09-25/isd.meta.yml [hunk 7]
```diff
@@ -4,12 +4,33 @@
...
+        unit: countries
+        processing_level: minor
+        description_processing: |-
+          We assign each sovereign state to a region based on the mapping (using ISD/COW codes):
+
+            - Americas: 2-199
+            - Europe: 200-399, 2558, 3375
+            - Africa: 400-626, 4044-6257
+            - Middle East: 630-699, 6821-6845
+            - Asia and Oceania: 700-999, 7003-9210
```
Comment (paarriagadap) [inline etl/steps/data/garden/countries/2023-09-25/isd.meta.yml:30]: These ids don't coincide with the ones used in the script

**Review by paarriagadap — APPROVED** [review]
Looks good! I have similar comments here

**Comment by lucasrodes** [comment]
Thanks, Pablo! I've incorporated your feedback.

File: etl/steps/data/garden/countries/2023-09-25/isd.meta.yml (hunk 7 above)
Comment (lucasrodes) [reply etl/steps/data/garden/countries/2023-09-25/isd.meta.yml:30]: Thanks for the catch! This should match now (except for codes 75911, 85421, which are custom made so I decided not to include them here).

**Pull request merged**